nsqd/protocol_v2.go

```diff
@@ -796,6 +797,9 @@ func (p *protocolV2) PUB(client *clientV2, params [][]byte) ([]byte, error) {
 		return nil, protocol.NewFatalClientErr(err, "E_PUB_FAILED", "PUB failed "+err.Error())
 	}
 
+	p.ctx.nsqd.AddClient(client.ID, client)
```

Comment:
I think what mreiferson is suggesting is to keep a set of all clients (added to the set when they connect). Then, when generating publisher statistics, you can iterate over all clients and include them only if they have published any messages.